# Largest Product in a Grid
source: https://projecteuler.net/problem=11

## Problem Statement
Given a 20x20 matrix of numbers in string form, find the largest product of 4 adjacent (horizontal, vertical, or diagonal) numbers?

Matrix: 

08 02 22 97 38 15 00 40 00 75 04 05 07 78 52 12 50 77 91 08
49 49 99 40 17 81 18 57 60 87 17 40 98 43 69 48 04 56 62 00
81 49 31 73 55 79 14 29 93 71 40 67 53 88 30 03 49 13 36 65
52 70 95 23 04 60 11 42 69 24 68 56 01 32 56 71 37 02 36 91
22 31 16 71 51 67 63 89 41 92 36 54 22 40 40 28 66 33 13 80
24 47 32 60 99 03 45 02 44 75 33 53 78 36 84 20 35 17 12 50
32 98 81 28 64 23 67 10 26 38 40 67 59 54 70 66 18 38 64 70
67 26 20 68 02 62 12 20 95 63 94 39 63 08 40 91 66 49 94 21
24 55 58 05 66 73 99 26 97 17 78 78 96 83 14 88 34 89 63 72
21 36 23 09 75 00 76 44 20 45 35 14 00 61 33 97 34 31 33 95
78 17 53 28 22 75 31 67 15 94 03 80 04 62 16 14 09 53 56 92
16 39 05 42 96 35 31 47 55 58 88 24 00 17 54 24 36 29 85 57
86 56 00 48 35 71 89 07 05 44 44 37 44 60 21 58 51 54 17 58
19 80 81 68 05 94 47 69 28 73 92 13 86 52 17 77 04 89 55 40
04 52 08 83 97 35 99 16 07 97 57 32 16 26 26 79 33 27 98 66
88 36 68 87 57 62 20 72 03 46 33 67 46 55 12 32 63 93 53 69
04 42 16 73 38 25 39 11 24 94 72 18 08 46 29 32 40 62 76 36
20 69 36 41 72 30 23 88 34 62 99 69 82 67 59 85 74 04 36 16
20 73 35 29 78 31 90 01 74 31 49 71 48 86 81 16 23 57 05 54
01 70 54 71 83 51 54 69 16 92 33 48 61 43 52 01 89 19 67 48


## Planning
For brute force search, there are a total of 1258 cases
* Vertical: (20 + 1 - 4) * 20 
* Horizontal: (20 + 1 - 4) * 20
* Diagonal:
    * northwest to southeast - (20 + 1 - 4) * (20 + 1 - 4)
    * northeast to southwest - (20 + 1 - 4) * (20 + 1 - 4) 
  
Break matrix string into row lists by separating at each '\n'

Break each row into element lists by separating at each ' '

Convert double list of strings into double list of ints

Then do brute force index parsing

In [1]:
2 * ((20+1-4)*20) + 2 * ((20+1-4)*(20+1-4))

1258

## [work]


In [5]:
# import numpy as np
    # actually, not needed

In [2]:
matrix_string = """08 02 22 97 38 15 00 40 00 75 04 05 07 78 52 12 50 77 91 08
49 49 99 40 17 81 18 57 60 87 17 40 98 43 69 48 04 56 62 00
81 49 31 73 55 79 14 29 93 71 40 67 53 88 30 03 49 13 36 65
52 70 95 23 04 60 11 42 69 24 68 56 01 32 56 71 37 02 36 91
22 31 16 71 51 67 63 89 41 92 36 54 22 40 40 28 66 33 13 80
24 47 32 60 99 03 45 02 44 75 33 53 78 36 84 20 35 17 12 50
32 98 81 28 64 23 67 10 26 38 40 67 59 54 70 66 18 38 64 70
67 26 20 68 02 62 12 20 95 63 94 39 63 08 40 91 66 49 94 21
24 55 58 05 66 73 99 26 97 17 78 78 96 83 14 88 34 89 63 72
21 36 23 09 75 00 76 44 20 45 35 14 00 61 33 97 34 31 33 95
78 17 53 28 22 75 31 67 15 94 03 80 04 62 16 14 09 53 56 92
16 39 05 42 96 35 31 47 55 58 88 24 00 17 54 24 36 29 85 57
86 56 00 48 35 71 89 07 05 44 44 37 44 60 21 58 51 54 17 58
19 80 81 68 05 94 47 69 28 73 92 13 86 52 17 77 04 89 55 40
04 52 08 83 97 35 99 16 07 97 57 32 16 26 26 79 33 27 98 66
88 36 68 87 57 62 20 72 03 46 33 67 46 55 12 32 63 93 53 69
04 42 16 73 38 25 39 11 24 94 72 18 08 46 29 32 40 62 76 36
20 69 36 41 72 30 23 88 34 62 99 69 82 67 59 85 74 04 36 16
20 73 35 29 78 31 90 01 74 31 49 71 48 86 81 16 23 57 05 54
01 70 54 71 83 51 54 69 16 92 33 48 61 43 52 01 89 19 67 48"""

In [3]:
print(matrix_string)

08 02 22 97 38 15 00 40 00 75 04 05 07 78 52 12 50 77 91 08
49 49 99 40 17 81 18 57 60 87 17 40 98 43 69 48 04 56 62 00
81 49 31 73 55 79 14 29 93 71 40 67 53 88 30 03 49 13 36 65
52 70 95 23 04 60 11 42 69 24 68 56 01 32 56 71 37 02 36 91
22 31 16 71 51 67 63 89 41 92 36 54 22 40 40 28 66 33 13 80
24 47 32 60 99 03 45 02 44 75 33 53 78 36 84 20 35 17 12 50
32 98 81 28 64 23 67 10 26 38 40 67 59 54 70 66 18 38 64 70
67 26 20 68 02 62 12 20 95 63 94 39 63 08 40 91 66 49 94 21
24 55 58 05 66 73 99 26 97 17 78 78 96 83 14 88 34 89 63 72
21 36 23 09 75 00 76 44 20 45 35 14 00 61 33 97 34 31 33 95
78 17 53 28 22 75 31 67 15 94 03 80 04 62 16 14 09 53 56 92
16 39 05 42 96 35 31 47 55 58 88 24 00 17 54 24 36 29 85 57
86 56 00 48 35 71 89 07 05 44 44 37 44 60 21 58 51 54 17 58
19 80 81 68 05 94 47 69 28 73 92 13 86 52 17 77 04 89 55 40
04 52 08 83 97 35 99 16 07 97 57 32 16 26 26 79 33 27 98 66
88 36 68 87 57 62 20 72 03 46 33 67 46 55 12 32 63 93 53 69
04 42 16 73 38 25 39 11 24 94 72 18 08 4

In [4]:
# testing
"alice, bob, charlie".split(',')

['alice', ' bob', ' charlie']

In [5]:
row_list = matrix_string.split('\n')
print(row_list)

['08 02 22 97 38 15 00 40 00 75 04 05 07 78 52 12 50 77 91 08', '49 49 99 40 17 81 18 57 60 87 17 40 98 43 69 48 04 56 62 00', '81 49 31 73 55 79 14 29 93 71 40 67 53 88 30 03 49 13 36 65', '52 70 95 23 04 60 11 42 69 24 68 56 01 32 56 71 37 02 36 91', '22 31 16 71 51 67 63 89 41 92 36 54 22 40 40 28 66 33 13 80', '24 47 32 60 99 03 45 02 44 75 33 53 78 36 84 20 35 17 12 50', '32 98 81 28 64 23 67 10 26 38 40 67 59 54 70 66 18 38 64 70', '67 26 20 68 02 62 12 20 95 63 94 39 63 08 40 91 66 49 94 21', '24 55 58 05 66 73 99 26 97 17 78 78 96 83 14 88 34 89 63 72', '21 36 23 09 75 00 76 44 20 45 35 14 00 61 33 97 34 31 33 95', '78 17 53 28 22 75 31 67 15 94 03 80 04 62 16 14 09 53 56 92', '16 39 05 42 96 35 31 47 55 58 88 24 00 17 54 24 36 29 85 57', '86 56 00 48 35 71 89 07 05 44 44 37 44 60 21 58 51 54 17 58', '19 80 81 68 05 94 47 69 28 73 92 13 86 52 17 77 04 89 55 40', '04 52 08 83 97 35 99 16 07 97 57 32 16 26 26 79 33 27 98 66', '88 36 68 87 57 62 20 72 03 46 33 67 46 55 12 32 63 93

In [6]:
element_list = []
for row in row_list:
    element_list.append(row.split(' '))
print(element_list)

[['08', '02', '22', '97', '38', '15', '00', '40', '00', '75', '04', '05', '07', '78', '52', '12', '50', '77', '91', '08'], ['49', '49', '99', '40', '17', '81', '18', '57', '60', '87', '17', '40', '98', '43', '69', '48', '04', '56', '62', '00'], ['81', '49', '31', '73', '55', '79', '14', '29', '93', '71', '40', '67', '53', '88', '30', '03', '49', '13', '36', '65'], ['52', '70', '95', '23', '04', '60', '11', '42', '69', '24', '68', '56', '01', '32', '56', '71', '37', '02', '36', '91'], ['22', '31', '16', '71', '51', '67', '63', '89', '41', '92', '36', '54', '22', '40', '40', '28', '66', '33', '13', '80'], ['24', '47', '32', '60', '99', '03', '45', '02', '44', '75', '33', '53', '78', '36', '84', '20', '35', '17', '12', '50'], ['32', '98', '81', '28', '64', '23', '67', '10', '26', '38', '40', '67', '59', '54', '70', '66', '18', '38', '64', '70'], ['67', '26', '20', '68', '02', '62', '12', '20', '95', '63', '94', '39', '63', '08', '40', '91', '66', '49', '94', '21'], ['24', '55', '58', '05'

In [8]:
for i in range(len(element_list)):
    for j in range(len(element_list[0])):
        element_list[i][j] = int(element_list[i][j])
print(element_list)

[[8, 2, 22, 97, 38, 15, 0, 40, 0, 75, 4, 5, 7, 78, 52, 12, 50, 77, 91, 8], [49, 49, 99, 40, 17, 81, 18, 57, 60, 87, 17, 40, 98, 43, 69, 48, 4, 56, 62, 0], [81, 49, 31, 73, 55, 79, 14, 29, 93, 71, 40, 67, 53, 88, 30, 3, 49, 13, 36, 65], [52, 70, 95, 23, 4, 60, 11, 42, 69, 24, 68, 56, 1, 32, 56, 71, 37, 2, 36, 91], [22, 31, 16, 71, 51, 67, 63, 89, 41, 92, 36, 54, 22, 40, 40, 28, 66, 33, 13, 80], [24, 47, 32, 60, 99, 3, 45, 2, 44, 75, 33, 53, 78, 36, 84, 20, 35, 17, 12, 50], [32, 98, 81, 28, 64, 23, 67, 10, 26, 38, 40, 67, 59, 54, 70, 66, 18, 38, 64, 70], [67, 26, 20, 68, 2, 62, 12, 20, 95, 63, 94, 39, 63, 8, 40, 91, 66, 49, 94, 21], [24, 55, 58, 5, 66, 73, 99, 26, 97, 17, 78, 78, 96, 83, 14, 88, 34, 89, 63, 72], [21, 36, 23, 9, 75, 0, 76, 44, 20, 45, 35, 14, 0, 61, 33, 97, 34, 31, 33, 95], [78, 17, 53, 28, 22, 75, 31, 67, 15, 94, 3, 80, 4, 62, 16, 14, 9, 53, 56, 92], [16, 39, 5, 42, 96, 35, 31, 47, 55, 58, 88, 24, 0, 17, 54, 24, 36, 29, 85, 57], [86, 56, 0, 48, 35, 71, 89, 7, 5, 44, 44, 

In [10]:
# splicing test
splice = element_list[1:5][1:5]

print(splice)

[[81, 49, 31, 73, 55, 79, 14, 29, 93, 71, 40, 67, 53, 88, 30, 3, 49, 13, 36, 65], [52, 70, 95, 23, 4, 60, 11, 42, 69, 24, 68, 56, 1, 32, 56, 71, 37, 2, 36, 91], [22, 31, 16, 71, 51, 67, 63, 89, 41, 92, 36, 54, 22, 40, 40, 28, 66, 33, 13, 80]]


In [15]:
# functions for obtaining diagonals
def diag_nw_se(matrix, i, j, length):
    """
    matrix: 2D list
    i: row index of first element (top left)
    j: column index of first element (top left)
    length: number of elements in diagonal
    
    Returns list of elements in specified diagonal, with wrap-around indices
    
    """
    diag = []
    
    for a in range(length):
        diag.append(matrix[(i+a)%len(matrix)][(j+a)%len(matrix[0])])
    
    return diag
    
    

In [16]:
# test 
diag_nw_se(element_list, 0, 0, 10)

[8, 49, 31, 23, 51, 3, 67, 20, 97, 45]

In [18]:
# functions for obtaining diagonals
def diag_sw_ne(matrix, i, j, length):
    """
    matrix: 2D list
    i: row index of first element (top left)
    j: column index of first element (top left)
    length: number of elements in diagonal
    
    Returns list of elements in specified diagonal, with wrap-around indices
    
    """
    diag = []
    
    for a in range(length):
        diag.append(matrix[(i-a)%len(matrix)][(j+a)%len(matrix[0])])
    
    return diag

In [19]:
# testing
diag_sw_ne(element_list, 5, 0, 5)

[24, 31, 95, 73, 17]

### Greatest vertical product

In [22]:
gvp = 0

for j in range(len(element_list[0])):
    
    for i in range( len(element_list) + 1 - 4):
        
        product = 1
        
        for k in range(4):
            product *= element_list[i+k][j]
        
        if product > gvp: gvp = product
        
print(f"Greatest vertical product: {gvp}")

Greatest vertical product: 51267216


### Greatest horizontal product

In [23]:
ghp = 0

for i in range(len(element_list)):
    
    for j in range( len(element_list[0]) + 1 - 4):
        
        product = 1
        
        for k in range(4):
            product *= element_list[i][j+k]
        
        if product > ghp: ghp = product
        
print(f"Greatest vertical product: {ghp}")

Greatest vertical product: 48477312


### Greatest diagonal (nw to se) product

In [24]:
gdp_nwse = 0

for i in range(len(element_list)+1-4):
    for j in range(len(element_list)+1-4):
        
        product = 1
        
        for k in range(4):
            product *= element_list[i+k][j+k]
            
        if product > gdp_nwse: gdp_nwse = product
            
print(f"Greatest diag (nw to se) product: {gdp_nwse}")

Greatest diag (nw to se) product: 40304286


### Greatest diagonal (sw to ne) product

In [28]:
gdp_swne = 0

start_index = (0,0)

for i in range(3, len(element_list)):
    
#     product = 1
    
    for j in range(len(element_list)+1-4):
        
        product = 1
        
        for k in range(4):
            product *= element_list[i-k][j+k]
        if product > gdp_swne: 
            gdp_swne = product
            start_index = (i,j)
            
print(f"Greatest diag (sw to ne) product: {gdp_swne}")
print(f"Index of start: {start_index}")

Greatest diag (sw to ne) product: 70600674
Index of start: (15, 3)


## Solution
By manual comparison, the greatest product of four consecutive numbers is 70600674 = 87 x 97 x 94 x 89